mobilecoind-json/src/main.rs

```diff
@@ -0,0 +1,317 @@
+#[get("/entropy")]
+fn entropy(state: rocket::State<State>) -> Result<Json<JsonEntropyResponse>, String> {
+    let resp = state
+        .mobilecoind_api_client
+        .generate_entropy(&mc_mobilecoind_api::Empty::new())
+        .map_err(|err| format!("Failed getting entropy: {}", err))?;
+    Ok(Json(JsonEntropyResponse {
+        entropy: hex::encode(resp.entropy),
+    }))
+}
+
+#[derive(Deserialize, Default)]
+struct JsonMonitorRequest {
+    key: String,
+    start: u64,
+    number: u64,
+}
+
+#[derive(Serialize, Default)]
+struct JsonMonitorResponse {
+    monitor_id: String,
+}
+
+/// Creates a monitor. Data for the key and range is POSTed using the struct above.
+#[post("/create-monitor", format = "json", data = "<monitor>")]
+fn create_monitor(
+    state: rocket::State<State>,
+    monitor: Json<JsonMonitorRequest>,
+) -> Result<Json<JsonMonitorResponse>, String> {
+    let entropy =
+        hex::decode(&monitor.key).map_err(|err| format!("Failed to decode hex key: {}", err))?;
+
+    let mut req = mc_mobilecoind_api::GetAccountKeyRequest::new();
+    req.set_entropy(entropy.to_vec());
+
+    let mut resp = state
+        .mobilecoind_api_client
+        .get_account_key(&req)
+        .map_err(|err| format!("Failed getting account key for entropy: {}", err))?;
+
+    let account_key = resp.take_account_key();
+
+    let mut req = mc_mobilecoind_api::AddMonitorRequest::new();
+    req.set_account_key(account_key);
+    req.set_first_subaddress(monitor.start);
+    req.set_num_subaddresses(monitor.number);
+    req.set_first_block(0);
+
+    let monitor_response = state
+        .mobilecoind_api_client
+        .add_monitor(&req)
+        .map_err(|err| format!("Failed adding monitor: {}", err))?;
+
+    Ok(Json(JsonMonitorResponse {
+        monitor_id: hex::encode(monitor_response.monitor_id),
+    }))
+}
+
+#[derive(Serialize, Default)]
+struct JsonBalanceResponse {
+    balance: u64,
+}
+
+/// Balance check using a created monitor and subaddress index
+#[get("/monitors/<monitor_hex>/balance/<index>")]
+fn balance(
+    state: rocket::State<State>,
+    monitor_hex: String,
+    index: u64,
+) -> Result<Json<JsonBalanceResponse>, String> {
+    let monitor_id =
+        hex::decode(monitor_hex).map_err(|err| format!("Failed to decode monitor hex: {}", err))?;
+
+    let mut req = mc_mobilecoind_api::GetBalanceRequest::new();
+    req.set_monitor_id(monitor_id.clone());
+    req.set_subaddress_index(index);
+
+    let resp = state
+        .mobilecoind_api_client
+        .get_balance(&req)
+        .map_err(|err| format!("Failed getting balance: {}", err))?;
+    let balance = resp.get_balance();
+    Ok(Json(JsonBalanceResponse { balance }))
+}
+
+#[derive(Serialize, Default)]
+struct JsonRequestResponse {
+    request_code: String,
+}
+
+/// Generates a request code without a balance or memo
+/// TODO: also add a POST that includes balance and memo 
+#[get("/monitors/<monitor_hex>/request-code/<index>")]
```

Comment:
I believe this should always be a POST.

Reply:
It should be a post, but also `new request-code` is an verb on a subaddress.